### PR TITLE
DOCSP-41760 Add transactions page

### PR DESCRIPTION
**Pull request by stephmarie17**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-kotlin/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-41760>
Staging - <https://preview-mongodbstephmarie17.gatsbyjs.io/kotlin/docsp-41760-addTransactionsPage/fundamentals/transactions/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/examples/generated/TransactionsTest.snippet.transaction.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// Set up the MongoDB client and get the collection
+suspend fun performTransaction(client: MongoClient) {
+    client.startSession().use { session ->
+        session.startTransaction() // Start the transaction
```
Comment (mcmorisi) [inline source/examples/generated/TransactionsTest.snippet.transaction.kt:4]: Typically I've always put comments on their own lines above the line they're meant to describe/annotate
```suggestion
        // Start the transaction
        session.startTransaction()
```

File: source/examples/generated/TransactionsTest.snippet.transaction.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                inc(SavingsAccount::amount, -100)
+            )
+         
+            val checkingColl = database.getCollection<CheckingAccount>("checking_accounts")
+            checkingColl.findOneAndUpdate(
+                session,
+                CheckingAccount::accountId eq "9876",
+                inc(CheckingAccount::amount, 100)
+            )
+            // Commit the transaction
```
Comment (mcmorisi) [inline source/examples/generated/TransactionsTest.snippet.transaction.kt:21]: Nit
```suggestion
            )

            // Commit the transaction
```

File: source/examples/generated/TransactionsTest.snippet.transaction.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        } catch (error: Exception) {
+            println("An error occurred during the transaction: ${error.message}")
+            // Abort the transaction
+            session.abortTransaction() 
+        }
+    }
+}
+
+data class SavingsAccount(val accountId: String, val amount: Int)
+data class CheckingAccount(val accountId: String, val amount: Int)
```
Comment (mcmorisi) [inline source/examples/generated/TransactionsTest.snippet.transaction.kt:33]: I'd lift these class definitions out of this example and introduce them above the code example proper.

I'd say after the 3 item list, you can say something like "This example uses the following {+language+} data classes to model its documents:" and then have these two lines above the bulk of the example. See [here](https://preview-mongodbmcmorisi.gatsbyjs.io/kotlin-sync/master/read/retrieve/) for something similar.

File: source/examples/generated/TransactionsTest.snippet.transaction.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+fun main() = runBlocking {
+    val uri = "<connection string uri>"
+    val client = MongoClient.create(uri)
+    performTransaction(client)
+}
```
Comment (mcmorisi) [inline source/examples/generated/TransactionsTest.snippet.transaction.kt:39]: I don't think we need these – for more niche pages like this one I tend to only include lines of code that strictly relate to the topic at hand for code examples, as opposed to everything necessary to get the program to run.

This main function would be necessary to copy-paste the code, but to my eyes as a reader adds length to the example, and therefore more room for confusion.

I personally would only include the lines within the `performTransaction` function you've defined here in the example, and then pulling out the data classes you've defined into a separate includes block (elaborated on in another comment).

**Review by mcmorisi — CHANGES_REQUESTED** [review]
The copy looks good, just some notes on the code sample!

File: source/examples/generated/TransactionsTest.snippet.transaction.kt (hunk 3 above)
Comment (stephmarie17) [reply source/examples/generated/TransactionsTest.snippet.transaction.kt:33]: That makes sense. Will update!

File: source/fundamentals/transactions.txt [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+1. Create a session from the client using the ``startSession()`` method.
+#. Use the ``startTransaction()`` method to start a transaction.
+#. Update the specified documents, then use the ``commitTransaction()`` method if all
+   operations succeed, or ``abortTransaction()`` if any operations fail.
+
+This example uses the following {+language+} data classes to model its documents:
+
+.. literalinclude:: /examples/generated/TransactionsTest.snippet.data-class.kt
+   :language: kotlin
+
```
Comment (mcmorisi) [inline source/fundamentals/transactions.txt:103]: Apologies for multiple rounds of feedback here – now that I see this on the page I feel that we need some sort of buffer between this code block and the example. I think I'd lift this to the top of the section, before "The following example demonstrates how..."

**Review by mcmorisi — APPROVED** [review]
LGTM with one last suggestion!

File: source/examples/generated/TransactionsTest.snippet.data-class.kt [hunk 6]
```diff
@@ -0,0 +1,3 @@
+    data class SavingsAccount(val accountId: String, val amount: Int)
+    data class CheckingAccount(val accountId: String, val amount: Int)
+
```
[the remark this answers is not in the document]
Comment (stephmarie17) [reply source/examples/generated/TransactionsTest.snippet.data-class.kt]: Updated code example to use a streamlined single data class 👍

**Comment by stephmarie17** [comment]
Tagging @mongodb/dbx-java for technical review 🙇‍♀️

**Comment by jyemin** [comment]
Am I correct that this page doesn't exist yet for the Java sync driver?  I was looking to compare the two.

**Comment by stephmarie17** [comment]
> Am I correct that this page doesn't exist yet for the Java sync driver? I was looking to compare the two.

Correct, we're in the process of adding these pages. There are a few PRs for Java-related transactions pages in progress:
- [Java Sync](https://github.com/mongodb/docs-java/pull/569)
- [Java RS](https://github.com/mongodb/docs-java-rs/pull/74)
Then one for Kotin sync will be upcoming. There are some examples of existing Transaction pages for other drivers in the JIRA ticket if that is helpful!

**Review by jyemin — COMMENTED** [review]
Everything looks good but wondering whether we should document `com.mongodb.kotlin.client.ClientSession#withTransaction` here as well, or if that's going in a different PR.

**Comment by stephmarie17** [comment]
@jyemin thanks for taking a look! I didn't see `withTransaction` in the API docs so I wasn't sure if that was supported. Is that supported and is there more info someplace in the API docs that I missed?

One other question I have: once this PR is approved and published, to what version of Kotlin Coroutine should we backport the docs update? The docs versions are upcoming, 5.1, 5.0, 4.11, and 4.10.

**Comment by jyemin** [comment]
> I didn't see withTransaction in the API docs so I wasn't sure if that was supported. Is that supported and is there more info someplace in the API docs that I missed?

Oh, I thought this was a Kotlin Sync Driver page.  Are you sure we don't need some sort of call to `runBlocking` anywhere?

> One other question I have: once this PR is approved and published, to what version of Kotlin Coroutine should we backport the docs update? The docs versions are upcoming, 5.1, 5.0, 4.11, and 4.10.

It's up to docs team how far all these pages will be backported.

**Comment by stephmarie17** [comment]
> > I didn't see withTransaction in the API docs so I wasn't sure if that was supported. Is that supported and is there more info someplace in the API docs that I missed?
> 
> Oh, I thought this was a Kotlin Sync Driver page. Are you sure we don't need some sort of call to `runBlocking` anywhere?
> 
Will be working on the Kotlin Sync page soon! There's a call to `runBlocking` in the example file (examples/src/test/kotlin/TransactionsTest.kt) but the [snippet displayed in the docs](https://deploy-preview-167--docs-kotlin.netlify.app/fundamentals/transactions/#example) starts after that.

**Review by jyemin — APPROVED** [review]
LGTM

**Pull request merged**